### PR TITLE
KHWE-31: Extract table as separate component and make sortable

### PR DESCRIPTION
**Pull request by ghost**
screen shot 2018-11-05 at 5 40 05 pm

### CODE REVIEW

File: src/components/Table/Table.js [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+
+  handleOrderClick(e) {
+    const { data, initialData, orderTypes, currentOrderTypeIndex } = this.state;
+
+    this.state.columnName = this.getColumnName(e.target);
+    if (!this.state.columnName) {
+      return;
+    }
+
+    if (currentOrderTypeIndex === 0) {
```
Comment (lubovgribiniyk) [inline src/components/Table/Table.js]: Should we use constants for order type indexes instead of just numbers like 0,1,2?

File: package.json [hunk 2]
```diff
@@ -15,6 +15,7 @@
   "license": "ISC",
   "dependencies": {
     "firebase": "5.5.6",
+    "lodash": "^4.17.11",
```
Comment (ihor4568) [inline package.json]: please freeze deps

**Comment by ihor4568** [comment]
Could you please use this design for table sorting?

image

File: src/components/Table/Table.js [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+      );
+    }
+
+    this.mount();
+  }
+
+  querySelectors() {
+    const { mountPoint } = this;
+    this.tableHead = mountPoint.querySelector(".table__head");
+    this.iconButtonRipple = mountPoint.querySelectorAll(".mdc-icon-button");
```
Comment (ihor4568) [inline src/components/Table/Table.js]: please do not query material class names

File: src/components/Table/Table.js [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+    this.dotsMenu = mountPoint.querySelectorAll(".table__td_more");
+    this.orderIcon = mountPoint.querySelector(
```
Comment (ihor4568) [inline src/components/Table/Table.js]: this will be undefined - why do you need it?

File: src/components/Table/Table.js (hunk 1 above)
Comment (ihor4568) [reply src/components/Table/Table.js]: ++

**Comment by ihor4568** [comment]
Why there is no sorting by number (first column) ?

File: src/components/Table/Table.js (hunk 1 above)
Comment (ghost) [reply src/components/Table/Table.js]: Edited to compare using constants.

File: package.json (hunk 2 above)
Comment (ghost) [reply package.json]: done

File: src/components/Table/Table.js (hunk 1 above)
Comment (ghost) [reply src/components/Table/Table.js]: done

File: src/components/Table/Table.js (hunk 3 above)
Comment (ghost) [reply src/components/Table/Table.js]: done

File: src/components/Table/Table.js [hunk 5]
```diff
@@ -0,0 +1,136 @@
...
+      columnName: ""
+    };
+
+    this.handleOrderClick = this.handleOrderClick.bind(this);
+  }
+
+  fillObjectsWithNumbersAsIndices(array) {
+    const newArray = [];
+    const indices = Object.keys(array);
+    indices.forEach(index => {
```
Comment (ihor4568) [inline src/components/Table/Table.js]: really ? array.map ((val, idx) => ...)

File: src/components/Table/Table.html [hunk 6]
```diff
@@ -0,0 +1,74 @@
+<table class="table">
+  <thead class="table__head">
+    <tr>
+      <th class="table__th"></th>
+      <th class="table__th table__th_number">
+        <button class="mdc-button table__th-button">
+          No.
+          <svg data="number" aria-hidden="true" class="mdc-button__icon table__th-icon table__th-icon_number" xmlns="http://www.w3.org/2000/svg" width="24" height="24" viewBox="0 0 24 24">
```
Comment (ihor4568) [inline src/components/Table/Table.html]: please create a folder called `assets/images` and move this svg there. BTW - maybe there is such arrow in material icons?

File: src/components/Table/Table.js [hunk 7]
```diff
@@ -0,0 +1,136 @@
...
+
+import tableTemplate from "./Table.html";
+import { DotsMenuComponent } from "../DotsMenu/DotsMenu";
+
+const ORDER_TYPES = {
+  0: "INITIAL",
+  1: "ASC",
+  2: "DESC"
+};
+
```
Comment (ihor4568) [inline src/components/Table/Table.js]: const INITIAL_ORDER = 0
const ASC_ORDER = 1
const DESC_ORDER = 2

Comment (ihor4568) [reply src/components/Table/Table.js]: why do you need this text representation? numeric constants should be fine

Comment (ghost) [reply src/components/Table/Table.js]: if (ORDER_TYPES[currentOrderTypeIndex] === "INITIAL") {
      this.state.data = initialData;
    } else {
      this.state.data = _.orderBy(
        data,
        this.state.columnName,
        ORDER_TYPES[currentOrderTypeIndex].toLowerCase()
      );
    }
It is required for the logic to be worked. It contains order type which I pass in to the orderBy function as an argument. I cannot pass in a number as the argument.

Comment (ihor4568) [reply src/components/Table/Table.js]: const INITIAL_ORDER = 0

...

if (currentOrderTypeIndex === INITIAL_ORDER )

Comment (ihor4568) [reply src/components/Table/Table.js]: oh, I see, you need to pass smth to _.orderBy

Comment (ghost) [reply src/components/Table/Table.js]: Yeah

Comment (ihor4568) [reply src/components/Table/Table.js]: then, 

const INITIAL_ORDER = 0
const ASC_ORDER = 1
const DESC_ORDER = 2

const ORDER_TYPES = {
  [INITIAL_ORDER]: 'initial',
  [ASC_ORDER]: 'asc',
  [DESC_ORDER]: 'desc'
}

OR

you can add one more `if` and write 'asc'/'desc' directly inside _.orderBy

File: src/components/Table/Table.js (hunk 5 above)
Comment (ghost) [reply src/components/Table/Table.js]: forgot this simple way. done

File: src/components/Table/Table.js [hunk 8]
```diff
@@ -0,0 +1,137 @@
...
+      this.orderIcon.classList.add("table__th-icon_order_asc");
+    } else if (currentOrderTypeIndex === DESC_ORDER) {
+      this.orderIcon.classList.add("table__th-icon_order_desc");
+    }
+
+    this.changeCurrentOrderTypeIndex();
+  }
+
+  changeCurrentOrderTypeIndex() {
+    if (this.state.currentOrderTypeIndex === 2) {
```
Comment (ihor4568) [inline src/components/Table/Table.js]: please use constants here (DESC_ORDER etc) and everywhere

File: src/components/Table/Table.js [hunk 9]
```diff
@@ -0,0 +1,137 @@
+import _ from "lodash";
```
Comment (ihor4568) [inline src/components/Table/Table.js]: please call this component SongsTableComponent. name Table is too generic. Do not forget to rename css classnames and files

File: src/components/Table/Table.js [hunk 10]
```diff
@@ -0,0 +1,137 @@
...
+      currentOrderTypeIndex: 1,
+      columnName: ""
+    };
+
+    this.handleOrderClick = this.handleOrderClick.bind(this);
+  }
+
+  fillObjectsWithNumbersAsIndices(array) {
+    return array.map((item, index) =>
+      Object.assign({}, array[index], { number: index })
```
Comment (ihor4568) [inline src/components/Table/Table.js]: you could use object rest spread here instead assign

**Review by ihor4568 — APPROVED** [review]

**Pull request merged**